### ISSUE TITLE
Balance update issue

### ISSUE THREAD

**Issue by bijakatlykkex** [MetacoSA/QBitNinja#10]
Hello,

It seems sometimes when a transaction is sent the balances are not updated. For example consider the following transaction: http://web.archive.org/web/20160616192047/http://api.qbit.ninja/transactions/acf3e98c37513699549f4d5df8911130736e3124ade4a1b71822ff7c663ce27c?colored=true the output 5 of transaction 184ef9cb8d34d62a9878bffbd25cb4473e70033913e793a9d567801b53b621c6 is spent in the captured transaction but this output it is still reported as unspent: http://web.archive.org/web/20160616191930/http://api.qbit.ninja/balances/38DdqhVuqb36jjmxTbvBkiFPXKA8EmW9Ly?unspentonly=true&colored=true

The problem seems to gradually be corrected, when time passes, since I had issues which seems similar to this from some days ago, but now they do not exist as I am investigating now.

I use QBitNinja.Listner.Console for updating.

**Comment by NicolasDorier** [comment]
thanks, how much time did it took for not being updated ?
I'm wondering if my listener is late on indexing.

**Comment by bijakatlykkex** [comment]
I am not sure, this problem has been reported to me as another problem occurring somewhere else which has happened consequently for some time, with the current root.

When I retried the operations, I think when something around 2 days has passed it was solved.

At the time of writing the reported address, still has the problem.

If I could find a narrower interval, I will inform you.

**Comment by NicolasDorier** [comment]
The is a possibility: The transaction does not reach my node for some reason, then when it get mined it is indexed. Is it possible ?

**Comment by bijakatlykkex** [comment]
The problem seems to be although  the transaction has been indexed, but the balance referencing that transaction has not been updated. So it seems the transaction reaches to the node.

The problem is also not only with api.qbit.ninja, at first it was seen in our own hosted copy of qbit.ninja and I wanted to report, I checked with api.qbit.ninja and I saw the problem is also present in api.qbit.ninja

**Comment by NicolasDorier** [comment]
is it only for one address ? maybe my instance get late on indexing.

**Comment by bijakatlykkex** [comment]
I have not another instance yet to proof. And it is not just your instance, ours was the same.

**Comment by bijakatlykkex** [comment]
One interesting thing is that the address 38DdqhVuqb36jjmxTbvBkiFPXKA8EmW9Ly has large number outputs either spent or unspent: https://btc-ninja.azurewebsites.net/balances/38DdqhVuqb36jjmxTbvBkiFPXKA8EmW9Ly , this may or may not be correlated.

**Comment by NicolasDorier** [comment]
I don't think this is the problem. I tested qbitninja on more than 10000 transaction and it worked fine.

What can happen is that the transaction get indexed when the balances are in process of getting indexed, and there is some time of lag between both. (should not take more than 1min though)

How much time were spent between when the balance was updated and when the transaction was indexed ?

**Comment by bijakatlykkex** [comment]
The balance has not been updated yet. Transaction  184ef9cb8d34d62a9878bffbd25cb4473e70033913e793a9d567801b53b621c6 output 5 is still reported unspent in 38DdqhVuqb36jjmxTbvBkiFPXKA8EmW9Ly although it is being spent in transaction acf3e98c37513699549f4d5df8911130736e3124ade4a1b71822ff7c663ce27c and relay time is Wednesday, June 15th 2016 as blocktrail says and still unconfirmed.

**Comment by bijakatlykkex** [comment]
Can it now be because more than timeout amount has passed and transaction has not get confirmation?

**Comment by NicolasDorier** [comment]
haa yes indeed, I am using a timeout if a transaction is unconf for too long. (I think it is something like 2H)

**Comment by bijakatlykkex** [comment]
Thanks. Is there any way to manually force such timed out balances to be reindexed?

**Comment by NicolasDorier** [comment]
well, one of the solution I can do is to add a parameter to the method which take the timeout, instead of hardcoding it.
I can also allow to configure the default one in app config.
But truth to be told, I think it is better to forget an operation in a balance that is not confirmed for long.

**Comment by bijakatlykkex** [comment]
Thanks. I have updated blockexplorer and updated the expiration time to 96 hours - 4 days from 6.0 hours (tomorrow at the time of writing): https://github.com/MetacoSA/QBitNinja/commit/919a17a7cb517971fcfebe1f1f1c7fe59c234db2

But I think the same problem of a spent output being reported as unspent in this 96 hours deployment, still persists.

Also the original problem happened on a deployment with 24 hour expiration. The transaction which its spending could not be retrieved correctly, seems to be tested about 2-3 hours of being broadcasted, the transaction indexed but the unspent outputs of the address did not reflect this in 2-3 hours period.

I may be wrong in reporting this, although now transaction has been mined and the problem not reproducible.

**Comment by NicolasDorier** [comment]
I don't understand. You updated to 4 days and still had the problem after 3 H ?

**Comment by bijakatlykkex** [comment]
I mean since from the initial event it has been passed more that 24 hours (transaction broadcasting happened on 15th), I wanted to check if I increase the expiration so it has not passed,as of today, will the problem of unspent outputs be corrected or not.

I increased to 96 hours and once as I checked the spent output was still being reported as unspent. I wanted to dig deeper and in the same time by accident the first confirmation came and everything was corrected, so no more investigation possible.
